ragen/llm_agent/es_manager.py

```
@@ -132,38 +207,68 @@ def _expand_seed(seed: int):
+        for entry in envs:
+            tag = entry['tag']
+            if tag not in tag_to_envs:
+                tag_to_envs[tag] = []
+            tag_to_envs[tag].append(entry)
+        
+        # Prepare data indices for each environment (sampling without replacement per tag)
+        env_data_indices = {}  # env_id -> data_idx or None
+        
+        # Sample without replacement for each tag
+        with all_seed(seed):
+            for tag, tag_envs in tag_to_envs.items():
```

Ruff (0.14.13)

Loop control variable `tag` not used within loop body

(B007)